Performance & Metrics changes (carries #56)

lornasong opened the pull request:
**Note: this PR carries on https://github.com/hashicorp/consul-esm/pull/56**

Changes that Bloomberg have been using in production to support 3,000+
nodes with on 100 ESM's.

 - Adds metrics for health checks and performance
 - Adds support for > 64 ESM Instances
 - Improvements to allow a checks to run ever 1s safely
 - Improvements to prevent spurious node status updates
 - Improvements to allow disabling node coordinate updates
 - Fix an issue where check updates can fail to propagate to ESM Followers
 - Extra Logging
 - Exposed new config options
 - Adds ability to skip node coordinate updates if they are small (reducing spurious updates)
 - Updated to go mod for vendored dependencies

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul-esm?pullRequest=58)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul-esm?pullRequest=58)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul-esm?pullRequest=58).</sub>

lornasong commented:
Notes on rebase and resolving merge conflicts. Below are files that were marked as conflicted, description of the conflict, and how I resolved them. Most notable is check.go

**check.go**
updateCheckHttp()
- Interval changed type. Updated to new type. `definition.Interval` => `definition.IntervalDuration` line 83
- Timeout changed type. Updated to new type. `definition.Timeout` => `definition.TimeoutDuration` line 84
- TLSSkipVerify changed to TLSClientConfig. Update to new field. line 86 & 87
- New check on check's status:
   - Added line `check.Status == latestCheck.Status` line 93
   - Added new parameter `latestCheck` in method signature `updateCheckHttp(latestCheck *api.HealthCheck ` line 76

updateCheckTCP()
- Similar to updateCheckHttp()
    - Interval type change: line 134
    - Timeout type change: line 135
    - Add status check: line 142
    - Pass new parameter for status check: line 129

UpdateChecks()
- Interval type change: line 201 & 202 & 217
- Pass in new parameter to updateCheckHttp and updateCheckTcp

**check_test.go**

Line above changes had changed. Kept new line:
`log.New(os.Stdout, "", 0)` => `logger := log.New(LOGOUT, "", 0)`

**config_test.go**

Line below new import had been removed. Removed: `"github.com/pascaldekloe/goe/verify"`

**go.mod/go.sum**
- Accepted incoming go.mod with the exception of consul and serf where I used the master version.
    - consul: v1.4.1-0.20190122193545-2abc48c1eb59 =>  v1.6.1
    - serf: v0.8.2-0.20180110034851-2d88a6b2e9b8 => v0.8.4
- Did a `go mod tidy`

**leader.go**
commitOps()
- Comment on method below commitOps(), computeWatchedNodes, was reformatted. Kept new formatting.

computeWatchedNodes()
- Formatting in a deleted block of code had changed. Kept the delete. Block: `success, results, _, err := a.client.KV().Txn(ops, nil)`

mikeyyuen commented:
@lornasong these changes look good, I pushed this rebase on to the original rebase https://github.com/hashicorp/consul-esm/pull/56 which should now have a signed CLA.

lornasong commented:
@mikeyyuan, thanks for reviewing and pushing the changes! I’ll close this PR.
